lib/HTML/Renderers/TableNodeRenderer.php

```diff
@@ -27,45 +27,27 @@ public function __construct(TableNode $tableNode, TemplateRenderer $templateRend
     public function render() : string
     {
         $headers = $this->tableNode->getHeaders();
-        $data    = $this->tableNode->getData();
+        $rows    = $this->tableNode->getData();
 
-        $tableHeader = [];
-        $tableRows   = [];
+        $tableHeaderRows = [];
 
-        if (count($headers) !== 0) {
-            foreach ($headers as $k => $isHeader) {
-                if (! isset($data[$k])) {
-                    continue;
-                }
-
-                /** @var SpanNode $col */
-                foreach ($data[$k] as $col) {
-                    $tableHeader[] = $col->render();
-                }
-
-                unset($data[$k]);
-            }
-        }
-
-        foreach ($data as $k => $row) {
-            if ($row === []) {
+        foreach ($headers as $k => $isHeader) {
+            if ($isHeader === false) {
                 continue;
             }
 
-            $tableRow = [];
-
-            /** @var SpanNode $col */
-            foreach ($row as $col) {
-                $tableRow[] = $col->render();
+            if (! isset($rows[$k])) {
+                throw new LogicException(sprintf('Row "%d" should be a header, but that row does not exist.', $k));
             }
 
-            $tableRows[] = $tableRow;
+            $tableHeaderRows[] = $rows[$k];
+            unset($rows[$k]);
         }
 
         return $this->templateRenderer->render('table.html.twig', [
             'tableNode' => $this->tableNode,
-            'tableHeader' => $tableHeader,
-            'tableRows' => $tableRows,
+            'tableHeaderRows' => $tableHeaderRows,
+            'tableRows' => $rows,
         ]);
```

Comment:
Because `$this->tableNode->getData();` now returns `TableRow[]` instead of some crazy array, I'm passing that straight to the template and dealing with it there.